platform/linux-dpdk/odp_crypto.c

```diff
@@ -868,35 +926,36 @@ int odp_crypto_session_create(odp_crypto_session_param_t *param,
 		cipher_xform.cipher.op = RTE_CRYPTO_CIPHER_OP_ENCRYPT;
 		auth_xform.auth.op = RTE_CRYPTO_AUTH_OP_GENERATE;
 		first_xform = &cipher_xform;
-		first_xform->next = &auth_xform;
+		if (auth_xform.auth.algo == RTE_CRYPTO_AUTH_NULL) {
+			first_xform->next = NULL;
+			res = CRYPTO_CHAIN_ONLY_CIPHER;
+		} else if (cipher_xform.cipher.algo == RTE_CRYPTO_CIPHER_NULL) {
+			first_xform = &auth_xform;
+			first_xform->next = NULL;
+			res = CRYPTO_CHAIN_ONLY_AUTH;
+		} else {
+			first_xform->next = &auth_xform;
+			res = CRYPTO_CHAIN_CIPHER_AUTH;
+		}
 	} else {
 		cipher_xform.cipher.op = RTE_CRYPTO_CIPHER_OP_DECRYPT;
 		auth_xform.auth.op = RTE_CRYPTO_AUTH_OP_VERIFY;
 		first_xform = &auth_xform;
-		first_xform->next = &cipher_xform;
-	}
-
-	rc = cipher_alg_odp_to_rte(param->cipher_alg, &cipher_xform);
-
-	/* Check result */
-	if (rc) {
-		*status = ODP_CRYPTO_SES_CREATE_ERR_INV_CIPHER;
-		return -1;
-	}
-
-	rc = auth_alg_odp_to_rte(param->auth_alg, &auth_xform);
-
-	/* Check result */
-	if (rc) {
-		*status = ODP_CRYPTO_SES_CREATE_ERR_INV_AUTH;
-		/* remove the crypto_session_entry_t */
-		memset(entry, 0, sizeof(*entry));
-		free_session(entry);
-		return -1;
+		if (cipher_xform.cipher.algo == RTE_CRYPTO_CIPHER_NULL) {
+			first_xform->next = NULL;
+			res = CRYPTO_CHAIN_ONLY_AUTH;
+		} else if (auth_xform.auth.algo == RTE_CRYPTO_AUTH_NULL) {
+			first_xform = &cipher_xform;
+			first_xform->next = NULL;
+			res = CRYPTO_CHAIN_ONLY_CIPHER;
+		} else {
+			first_xform->next = &cipher_xform;
+			res = CRYPTO_CHAIN_AUTH_CIPHER;
+		}
```

Comment:
I'd suggest to use ALG_NULL checks directly, rather than introduce chain order.

Reply:
Not sure if I have fallowed. I introduced Chain order because dpdk PMD's doesn't have support for combination of NULL + Valid algo. So linux-dpdk implementation converts the NULL also to chain order as if it cipher only & auth only.

Reply:
I suggest to factor out cipher and auth checks as separate functions and apply them only if corresponding algorithm is not NULL. Code would be simpler.